Mathlib/Control/Monad/Free.lean

```diff
@@ -0,0 +1,601 @@
+    | liftBind op cont ih =>
+      simp_all [← map_eq_functorMap, ← liftBind_eq_lift_bind, map, ih]
+
+instance : Monad (FreeM F) where
+
+instance : LawfulMonad (FreeM F) := LawfulMonad.mk'
+  (bind_pure_comp := fun f x => by {
+    induction x with
+    | pure a => rfl
+    | liftBind op cont ih =>
+      simp only [FreeM.bind, ← bind_eq_bindBind, ← map_eq_functorMap, ← pure_eq_purePure, map] at *
+      simp only [ih]
+  })
+  (id_map := id_map)
+  (pure_bind := fun x f => rfl)
+  (bind_assoc := FreeM.bind_assoc)
+
+/--
+Interpret a `FreeM f` computation into any monad `m` by providing an interpretation
+function for the effect signature `f`.
+
+This function defines the *canonical interpreter* from the free monad `FreeM f` into the target
+monad `m`. It is the unique monad morphism that extends the effect handler
+`interp : ∀ {β}, F β → M β` via the universal property of `FreeM`.
+-/
+protected def liftM {M : Type u → Type w} [Monad M] {α : Type u} :
+    FreeM F α → ({β : Type u} → F β → M β) → M α
+  | .pure a, _ => pure a
+  | .liftBind op cont, interp => interp op >>= fun result => (cont result).liftM interp
+
+@[simp]
+lemma liftM_pure {M : Type u → Type w} [Monad M] {α : Type u}
+    (interp : {β : Type u} → F β → M β) (a : α) :
+    (pure a : FreeM F α).liftM interp = pure a := by rfl
+
+lemma liftM_liftBind {M : Type u → Type w} [Monad M] {α β : Type u}
+    (interp : {γ : Type u} → F γ → M γ) (op : F β) (cont : β → FreeM F α) :
+    (liftBind op cont).liftM interp = (do let b ← interp op; (cont b).liftM interp) := by
+  simp [FreeM.liftM]
+
+@[simp]
+lemma liftM_bind {M : Type u → Type w} [Monad M] [LawfulMonad M] {α β : Type u}
+    (interp : {β : Type u} → F β → M β) (x : FreeM F α) (f : α → FreeM F β) :
+    (x >>= f : FreeM F β).liftM interp = (do let a ← x.liftM interp; (f a).liftM interp) := by
+  induction x generalizing f with
+  | pure a => simp [pure_bind]
+  | liftBind op cont ih =>
+    simp_rw [← bind_eq_bindBind] at *
+    rw [FreeM.bind, liftM_liftBind, liftM_liftBind, bind_assoc]
+    simp_rw [ih]
+
+@[simp]
+lemma liftM_lift {M : Type u → Type w} [Monad M] [LawfulMonad M] {α : Type u}
+    (interp : {β : Type u} → F β → M β) (op : F α) :
+    (lift op).liftM interp = interp op := by
+  simp_rw [lift, liftM_liftBind, liftM_pure, bind_pure]
+
+/--
+A predicate stating that `g : FreeM F α → M α` is an interpreter for the effect
+handler `f : ∀ {α}, F α → M α`.
+
+This means that `g` is a monad morphism from the free monad `FreeM F` to the
+monad `M`, and that it extends the interpretation of individual operations
+given by `f`.
+
+Formally, `g` satisfies the two equations:
+- `g (pure a) = pure a`
+- `g (liftBind op k) = f op >>= fun x => g (k x)`
+-/
+def ExtendsHandler
+    {M : Type u → Type w} [Monad M] {α : Type u}
+    (f : {ι : Type u} → F ι → M ι)
+    (g : FreeM F α → M α) : Prop :=
+  (∀ a, g (pure a) = pure a) ∧
+  (∀ {ι} (op : F ι) (k : ι → FreeM F α),
+    g (liftBind op k) = f op >>= fun x => g (k x))
+
```

Comment:
This probably works better as a two-field structure, with `apply_pure` and `apply_liftBind` as fields.